spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java

```diff
@@ -166,12 +169,39 @@ protected void set(InternalRow struct, int pos, Object value) {
 
   static class UnionReader implements OrcValueReader<Object> {
     private final OrcValueReader[] readers;
+    private final Type expected;
+    private int[] projectedFieldIdsToIdxInReturnedRow;
+    private boolean isTagFieldProjected;
+    private int numOfFieldsInReturnedRow;
 
-    private UnionReader(List<OrcValueReader<?>> readers) {
+    private UnionReader(List<OrcValueReader<?>> readers, Type expected) {
       this.readers = new OrcValueReader[readers.size()];
       for (int i = 0; i < this.readers.length; i += 1) {
         this.readers[i] = readers.get(i);
       }
+      this.expected = expected;
+
+      if (this.readers.length > 1) {
+        // Creating an integer array to track the mapping between the index of fields to be projected
+        // and the index of the value for the field stored in the returned row,
+        // if the value for a field equals to -1, it means the value of this field should not be stored
```

Comment:
can we use something apart from '-1' since we use that constant as a fieldId for invalid dummy fields. 
INT_MIN?

Reply:
Good point, done